File: sdks/php/src/Api/ContactProperties.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\ContactPropertyPageResponse;
use Apollo\Deploy\Signal\Sdk\ContactPropertyResponse;
use Apollo\Deploy\Signal\Sdk\CreateContactPropertyBody;
use Apollo\Deploy\Signal\Sdk\UpdateContactPropertyBody;
use Psr\Http\Message\ResponseInterface;

/** ContactPropertiesAPI domain operations. */
class ContactProperties
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param string $projectId
     * @return ContactPropertyPageResponse
     * @throws SDKError
     */
    public function listContactProperties(
        string $projectId,
    ): ContactPropertyPageResponse
    {
        $path = '/v1/projects/${projectId}/contact-properties';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactPropertyPageResponse::class, 'fromArray')) {
            return ContactPropertyPageResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $propertyId
     * @return ContactPropertyResponse
     * @throws SDKError
     */
    public function getContactProperty(
        string $projectId,
        string $propertyId,
    ): ContactPropertyResponse
    {
        $path = '/v1/projects/${projectId}/contact-properties/${propertyId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${propertyId}', rawurlencode(Transport::wireValue($propertyId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactPropertyResponse::class, 'fromArray')) {
            return ContactPropertyResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param CreateContactPropertyBody $input
     * @return ContactPropertyResponse
     * @throws SDKError
     */
    public function createContactProperty(
        string $projectId,
        CreateContactPropertyBody $input,
    ): ContactPropertyResponse
    {
        $path = '/v1/projects/${projectId}/contact-properties';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactPropertyResponse::class, 'fromArray')) {
            return ContactPropertyResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $propertyId
     * @param UpdateContactPropertyBody $input
     * @return ContactPropertyResponse
     * @throws SDKError
     */
    public function updateContactProperty(
        string $projectId,
        string $propertyId,
        UpdateContactPropertyBody $input,
    ): ContactPropertyResponse
    {
        $path = '/v1/projects/${projectId}/contact-properties/${propertyId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${propertyId}', rawurlencode(Transport::wireValue($propertyId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'PATCH',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactPropertyResponse::class, 'fromArray')) {
            return ContactPropertyResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $propertyId
     * @return void
     * @throws SDKError
     */
    public function deleteContactProperty(
        string $projectId,
        string $propertyId,
    ): void
    {
        $path = '/v1/projects/${projectId}/contact-properties/${propertyId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${propertyId}', rawurlencode(Transport::wireValue($propertyId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/contact_properties.rs
use crate::error::SdkError;
use crate::transport::Transport;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ ContactPropertyPageResponse, ContactPropertyResponse, CreateContactPropertyBody, UpdateContactPropertyBody };

/// ContactPropertiesAPI API operations.
pub struct ContactPropertiesAPI {
    transport: Transport,
}

impl ContactPropertiesAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn list_contact_properties(
        &self,
        project_id: String,
    ) -> Result<ContactPropertyPageResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contact-properties".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactPropertyPageResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_contact_property(
        &self,
        project_id: String,
        property_id: String,
    ) -> Result<ContactPropertyResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contact-properties/${propertyId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${propertyId}", &encode_path(&property_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactPropertyResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn create_contact_property(
        &self,
        project_id: String,
        body: &CreateContactPropertyBody,
    ) -> Result<ContactPropertyResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contact-properties".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactPropertyResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn update_contact_property(
        &self,
        project_id: String,
        property_id: String,
        body: &UpdateContactPropertyBody,
    ) -> Result<ContactPropertyResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contact-properties/${propertyId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${propertyId}", &encode_path(&property_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactPropertyResponse>(
            Method::PATCH,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn delete_contact_property(
        &self,
        project_id: String,
        property_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/contact-properties/${propertyId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${propertyId}", &encode_path(&property_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
